src/main/kotlin/cacophony/automata/Determinization.kt

```diff
@@ -0,0 +1,87 @@
+package cacophony.automata
+
+import kotlin.collections.mutableListOf
+import kotlin.collections.mutableMapOf
+import kotlin.collections.setOf
+
+public fun <StateType> determinize(nfa: NFA<StateType>): DFA<Int> =
+    determinize(
+        nfa,
+        nfa.getProductions().keys.fold(mutableSetOf()) { charset, (_, char) ->
+            charset.add(char)
+            charset
+        },
+    )
+
+private fun <StateType> determinize(
+    nfa: NFA<StateType>,
+    charset: Iterable<Char>,
+): DFA<Int> {
+    val startingState = setOf(nfa.getStartingState())
+    var list = mutableListOf(startingState)
+    var dfaProductions = mutableMapOf<Set<StateType>, MutableMap<Char, Set<StateType>>>()
+
+    while (!list.isEmpty()) {
```

Comment:
Are we missing pushing to list? And the list could also be some queue object.

Reply:
Very interesting BFS implementation indeed